extensions/smallrye-jwt/deployment/src/main/java/io/quarkus/smallrye/jwt/deployment/SmallRyeJwtProcessor.java

```diff
@@ -161,14 +165,35 @@ void registerOptionalClaimProducer(BeanRegistrationPhaseBuildItem beanRegistrati
                 continue;
             }
             AnnotationInstance claimQualifier = injectionPoint.getRequiredQualifier(CLAIM_NAME);
-            if (claimQualifier != null && injectionPoint.getType().name().equals(DotNames.PROVIDER)) {
-                // Classes from jakarta.json are handled specially
+            if (claimQualifier != null) {
+                boolean providerType = injectionPoint.getType().name().equals(DotNames.PROVIDER);
+                boolean instanceType = injectionPoint.getType().name().equals(DotNames.INSTANCE);
+
                 Type actualType = injectionPoint.getRequiredType();
-                if (actualType.name().equals(DotNames.OPTIONAL) && !actualType.name().toString()
-                        .startsWith("jakarta.json")) {
-                    additionalTypes.add(actualType);
+
+                Optional<BeanInfo> bean = injectionPoint.getTargetBean();
+                if (bean.isPresent()) {
+                    DotName scope = bean.get().getScope().getDotName();
+                    if ((DotNames.SINGLETON.equals(scope) || DotNames.APPLICATION_SCOPED.equals(scope))
+                            && (!providerType && !instanceType && !CLAIM_VALUE_NAME.equals(actualType.name()))) {
+                        String error = String.format(
+                                "%s type can not be used to represent JWT claims in @Singleton or @ApplicationScoped beans"
+                                        + ", make the bean @RequestScoped or wrap this type with org.eclipse.microprofile.jwt.ClaimValue"
+                                        + " or jakarta.inject.Provider or jakarta.enterprise.inject.Instance",
+                                actualType.name());
+                        throw new IllegalStateException(error);
+                    }
+                }
+
+                if (providerType) {
+                    // Classes from jakarta.json are handled specially
+                    if (actualType.name().equals(DotNames.OPTIONAL) && !actualType.name().toString()
```

Comment:
This is not a new code, but you are moving it, so it should also be part of the review. If `actualType.name().equals(DotNames.OPTIONAL)` which is `java.util.Optional`, then `toString()` for `DotName` is `java.util.Optional`.  So `&& !actualType.name().toString().startsWith("jakarta.json")` is implicitly true for `Optional`.

Reply:
@michalvavrik I honestly don't know what this code does - Martin @mkouba, do you recall its purpose ?

Reply:
Oh I see, @michalvavrik, we can just remove this extra boolean check, thanks for spotting, and yes, recall now what this code does after all :-), there is some special handling for the Optional type

Reply:
@michalvavrik @mkouba I've removed the redundant `jakarta.json` check - it might've been needed in some form in the early days, but all passes now without it. Dropping the Optional check causes failures so I'm keeping it - in any case this is a clean up not related to the PR itself